File: packages/cli/src/scanner.ts
import {
  RUNTIME_VERSION,
  SCHEMA_VERSION,
  summarize,
  type AccessibilityReport,
  type Profile,
  type ReportItem,
  type SourceLocation,
} from './report';

export interface ScanOptions {
  target: string;
  profile?: Profile;
  runtime?: 'node' | 'python';
}

interface TagToken {
  name: string;
  attributes: Record<string, string>;
  index: number;
  location: SourceLocation;
}

function lineAt(source: string, index: number): number {
  return source.slice(0, index).split('\n').length;
}

function locationAt(source: string, index: number, excerpt: string): SourceLocation {
  const lineStart = source.lastIndexOf('\n', index - 1) + 1;
  return {
    line: lineAt(source, index),
    column: index - lineStart + 1,
    excerpt: excerpt.replace(/\s+/g, ' ').trim().slice(0, 160),
  };
}

function parseAttributes(raw: string): Record<string, string> {
  const attributes: Record<string, string> = {};
  const pattern = /([^\s=/>]+)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s>]+)))?/g;
  for (const match of raw.matchAll(pattern)) {
    attributes[match[1].toLowerCase()] = match[2] ?? match[3] ?? match[4] ?? '';
  }
  return attributes;
}

function tagsIn(source: string): TagToken[] {
  const tags: TagToken[] = [];
  const pattern = /<([a-zA-Z][\w:-]*)(\s[^<>]*?)?\s*\/?>/g;
  for (const match of source.matchAll(pattern)) {
    tags.push({
      name: match[1].toLowerCase(),
      attributes: parseAttributes(match[2] ?? ''),
      index: match.index,
      location: locationAt(source, match.index, match[0]),
    });
  }
  return tags;
}

function item(
  values: Pick<
    ReportItem,
    | 'ruleId'
    | 'classification'
    | 'certainty'
    | 'severity'
    | 'evidence'
    | 'message'
    | 'remediation'
    | 'verification'
  > & { wcag?: string[]; location?: SourceLocation | null },
): ReportItem {
  return { wcag: [], location: null, ...values };
}

function strippedText(fragment: string): string {
  return fragment
    .replace(/<[^>]+>/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function elementFragments(source: string, name: string): Array<{ text: string; index: number }> {
  const pattern = new RegExp(`<${name}\\b[^>]*>([\\s\\S]*?)<\\/${name}>`, 'gi');
  return Array.from(source.matchAll(pattern), (match) => ({
    text: strippedText(match[1]),
    index: match.index,
  }));
}

function manualBaseline(): ReportItem[] {
  return [
    item({
      ruleId: 'color-only-communication',
      classification: 'normative',
      certainty: 'manual',
      severity: 'info',
      evidence: 'human',
      message:
        'Confirm that color is not the only way information, status, or required action is conveyed.',
      wcag: ['1.4.1'],
      remediation:
        'Pair color with text, shape, pattern, position, or another programmatically available cue.',
      verification: 'Review each state visually and with color or custom styles disabled.',
    }),
    item({
      ruleId: 'contrast-rendered',
      classification: 'normative',
      certainty: 'manual',
      severity: 'info',
      evidence: 'runtime',
      message:
        'Measure rendered foreground and background colors, including states, images, gradients, and transparency.',
      wcag: ['1.4.3', '1.4.11'],
      remediation:
        'Adjust rendered colors that do not meet the applicable text or non-text threshold.',
      verification: 'Measure computed colors in the browser and inspect every interactive state.',
    }),
  ];
}

function cviChecks(): ReportItem[] {
  const shared = {
    classification: 'supplemental' as const,
    certainty: 'manual' as const,
    severity: 'info' as const,
    evidence: 'human' as const,
    wcag: [] as string[],
    location: null,
  };
  return [
    item({
      ...shared,
      ruleId: 'cvi-individual-profile',
      message:
        'Review the experience against the person’s individual CVI profile and preferred presentation.',
      remediation: 'Make presentation, contrast, spacing, field complexity, and pacing adjustable.',
      verification:
        'Evaluate with the person or a qualified specialist; no universal palette or ratio establishes CVI access.',
    }),
    item({
      ...shared,
      ruleId: 'cvi-visual-complexity',
      message: 'Review clutter, crowding, background complexity, and competing visual targets.',
      remediation: 'Reduce simultaneous visual demands and allow isolation of the current task.',
      verification:
        'Observe target recognition across realistic screens, not isolated components only.',
    }),
    item({
      ...shared,
      ruleId: 'cvi-fatigue-motion',
      message:
        'Review visual fatigue, latency, motion, and the time needed to find and interpret targets.',
      remediation: 'Provide pauses, stable layouts, reduced motion, and enough processing time.',
      verification: 'Test over a representative session with the person’s preferred pacing.',
    }),
    item({
      ...shared,
      ruleId: 'cvi-multisensory-alternatives',
      message:
        'Confirm that important information has usable nonvisual or multisensory alternatives.',
      remediation:
        'Provide meaningful speech, sound, touch, or text alternatives chosen with the person.',
      verification: 'Complete key tasks using the person’s preferred combination of senses.',
    }),
  ];
}

function switchChecks(): ReportItem[] {
  const shared = {
    classification: 'supplemental' as const,
    certainty: 'manual' as const,
    severity: 'info' as const,
    evidence: 'human' as const,
    wcag: [] as string[],
    location: null,
  };
  return [
    item({
      ...shared,
      ruleId: 'switch-control-verification',
      message: 'Complete the primary flow with the operating system’s Switch Control.',
      remediation:
        'Use native semantics, predictable focus order, and visible focus; remove unreachable controls.',
      verification:
        'Navigate, activate, dismiss, recover from errors, and complete the task using switch input only.',
    }),
    item({
      ...shared,
      ruleId: 'switch-timing-preferences',
      message:
        'Confirm scan speed, dwell, repeat filtering, and time limits can match the person’s needs.',
      remediation: 'Expose timing choices and avoid a universal dwell or repeat interval.',
      verification: 'Test slow and fast settings with the person’s switch configuration.',
    }),
    item({
      ...shared,
      ruleId: 'switch-simple-action',
      message:
        'Confirm complex pointer gestures have a single-switch or simple-action alternative.',
      wcag: ['2.5.1', '2.5.7'],
      remediation: 'Add controls for drag, path, multipoint, and pointer-only actions.',
      verification:
        'Complete every pointer-driven action using focus and a single activation command.',
    }),
  ];
}

/** Scan HTML source without claiming browser, assistive-technology, or human verification. */
export function scanSource(source: string, options: ScanOptions): AccessibilityReport {
  const tags = tagsIn(source);
  const findings: ReportItem[] = [];
  const manualChecks = manualBaseline();

  const html = tags.find((tag) => tag.name === 'html');
  if (!html?.attributes.lang?.trim()) {
    findings.push(
      item({
        ruleId: 'document-language',
        classification: 'normative',
        certainty: 'detected',
        severity: 'error',
        evidence: 'source',
        message: 'The document does not declare a default language.',
        wcag: ['3.1.1'],
        location: html?.location ?? { line: 1, column: 1 },
        remediation: 'Add a valid lang attribute to the html element.',
        verification:
          'Inspect the accessibility tree and confirm language changes are also marked.',
      }),
    );
  }

  const ids = new Map<string, TagToken>();
  for (const tag of tags) {
    const id = tag.attributes.id;
    if (!id) continue;
    const first = ids.get(id);
    if (first) {
      findings.push(
        item({
          ruleId: 'duplicate-id',
          classification: 'normative',
          certainty: 'detected',
          severity: 'error',
          evidence: 'source',
          message: `The id ${JSON.stringify(id)} is used more than once.`,
          wcag: ['1.3.1', '4.1.2'],
          location: tag.location,
          remediation: 'Give each relationship target a unique id and update its references.',
          verification: `Confirm every reference to ${JSON.stringify(id)} resolves to one intended element.`,
        }),
      );
    } else {
      ids.set(id, tag);
    }
  }

  for (const tag of tags.filter((candidate) => candidate.name === 'img')) {
    if (!Object.hasOwn(tag.attributes, 'alt')) {
      findings.push(
        item({
          ruleId: 'image-alt-missing',
          classification: 'normative',
          certainty: 'detected',
          severity: 'error',
          evidence: 'source',
          message: 'An image has no alt attribute.',
          wcag: ['1.1.1'],
          location: tag.location,
          remediation:
            'Add contextual alternative text, or alt="" when the image is intentionally decorative.',
          verification:
            'Review the image in context and confirm the accessible name communicates its purpose.',
        }),
      );
    } else if (tag.attributes.alt.trim() === '') {
      manualChecks.push(
        item({
          ruleId: 'image-alt-empty-context',
          classification: 'normative',
          certainty: 'manual',
          severity: 'info',
          evidence: 'human',
          message:
            'An image uses empty alternative text; that can be correct only when its content is redundant or decorative.',
          wcag: ['1.1.1'],
          location: tag.location,
          remediation:
            'Keep alt="" for decorative images; otherwise provide concise contextual alternative text.',
          verification: 'Review adjacent content and the task with images unavailable.',
        }),
      );
    }
  }

  const labelFors = new Set(
    tags
      .filter((tag) => tag.name === 'label' && tag.attributes.for)
      .map((tag) => tag.attributes.for),
  );
  for (const tag of tags.filter((candidate) =>
    ['input', 'select', 'textarea'].includes(candidate.name),
  )) {
    const type = (tag.attributes.type || 'text').toLowerCase();
    if (['hidden', 'button', 'submit', 'reset', 'image'].includes(type)) continue;
    const named =
      (tag.attributes.id && labelFors.has(tag.attributes.id)) ||
      tag.attributes['aria-label']?.trim() ||
      tag.attributes['aria-labelledby']?.trim() ||
      tag.attributes.title?.trim();
    if (!named) {
      findings.push(
        item({
          ruleId: 'form-label',
          classification: 'normative',
          certainty: 'potential',
          severity: 'error',
          evidence: 'source',
          message: 'A form control has no source-visible label or accessible-name reference.',
          wcag: ['1.3.1', '3.3.2', '4.1.2'],
          location: tag.location,
          remediation:
            'Associate a visible label with the control and preserve its accessible name.',
          verification:
            'Inspect the computed accessible name; account for labels created at runtime.',
        }),
      );
    }
  }

  for (const tag of tags) {
    const tabindex = tag.attributes.tabindex;
    if (tabindex !== undefined && Number.parseInt(tabindex, 10) > 0) {
      findings.push(
        item({
          ruleId: 'focus-positive-tabindex',
          classification: 'normative',
          certainty: 'potential',
          severity: 'warning',
          evidence: 'source',
          message: 'A positive tabindex can create a focus order that diverges from reading order.',
          wcag: ['2.4.3'],
          location: tag.location,
          remediation:
            'Use DOM order and tabindex="0" or native focusability instead of a positive value.',
          verification: 'Traverse the complete flow with Tab, Shift+Tab, and Switch Control.',
        }),
      );
    }
  }

  let previousHeading = 0;
  for (const tag of tags.filter((candidate) => /^h[1-6]$/.test(candidate.name))) {
    const level = Number(tag.name[1]);
    if (previousHeading && level > previousHeading + 1) {
      findings.push(
        item({
          ruleId: 'heading-order',
          classification: 'advisory',
          certainty: 'potential',
          severity: 'warning',
          evidence: 'source',
          message: `Heading level jumps from h${previousHeading} to h${level}.`,
          wcag: ['1.3.1', '2.4.6'],
          location: tag.location,
          remediation:
            'Use heading levels that reflect the content hierarchy without skipping a parent level.',
          verification:
            'Review the heading outline and confirm it matches the information structure.',
        }),
      );
    }
    previousHeading = level;
  }

  for (const link of elementFragments(source, 'a')) {
    if (!link.text || /^(click here|read more|learn more|more)$/i.test(link.text)) {
      findings.push(
        item({
          ruleId: 'link-purpose',
          classification: 'normative',
          certainty: 'potential',
          severity: 'warning',
          evidence: 'source',
          message: 'A link has empty or generic source-visible text.',
          wcag: ['2.4.4'],
          location: locationAt(source, link.index, source.slice(link.index, link.index + 120)),
          remediation: 'Write link text that identifies the destination or action in context.',
          verification: 'Review the computed accessible name and a links-only list.',
        }),
      );
    }
  }

  for (const table of source.matchAll(/<table\b[^>]*>([\s\S]*?)<\/table>/gi)) {
    if (!/<th\b/i.test(table[1])) {
      manualChecks.push(
        item({
          ruleId: 'table-headers-context',
          classification: 'normative',
          certainty: 'manual',
          severity: 'info',
          evidence: 'human',
          message:
            'A table has no source-visible header cells; determine whether it presents data or is only layout.',
          wcag: ['1.3.1'],
          location: locationAt(source, table.index, table[0].slice(0, 120)),
          remediation:
            'For data tables, add correctly scoped headers and a useful caption when needed.',
          verification: 'Navigate the table by row and column with a screen reader.',
        }),
      );
    }
  }

  for (const tag of tags) {
    const interactive =
      ['button', 'input', 'select', 'textarea'].includes(tag.name) ||
      (tag.name === 'a' && Boolean(tag.attributes.href)) ||
      Boolean(tag.attributes.role?.match(/^(button|link|checkbox|radio|switch|tab|menuitem)$/));
    if (!interactive) continue;
    const style = tag.attributes.style ?? '';
    const width = /(?:^|;)\s*width\s*:\s*([\d.]+)px/i.exec(style)?.[1];
    const height = /(?:^|;)\s*height\s*:\s*([\d.]+)px/i.exec(style)?.[1];
    if ((width && Number(width) < 24) || (height && Number(height) < 24)) {
      manualChecks.push(
        item({
          ruleId: 'target-size-spacing',
          classification: 'normative',
          certainty: 'manual',
          severity: 'info',
          evidence: 'runtime',
          message:
            'A source dimension is below 24 CSS pixels; rendered size, spacing, and WCAG exceptions still require review.',
          wcag: ['2.5.8'],
          location: tag.location,
          remediation:
            'Increase the rendered target or provide sufficient spacing or an applicable exception.',
          verification:
            'Measure rendered boxes and spacing at representative viewport sizes and zoom levels.',
        }),
      );
    }
  }

  for (const tag of tags) {
    const hasExecutableTiming =
      (['audio', 'video'].includes(tag.name) && Object.hasOwn(tag.attributes, 'autoplay')) ||
      (tag.name === 'meta' && tag.attributes['http-equiv']?.toLowerCase() === 'refresh');
    if (hasExecutableTiming) {
      manualChecks.push(
        item({
          ruleId: 'timing-media-control',
          classification: 'normative',
          certainty: 'manual',
          severity: 'info',
          evidence: 'runtime',
          message:
            'Source markup configures automatic media or refresh behavior that needs runtime timing review.',
          wcag: ['2.2.1', '2.2.2'],
          location: tag.location,
          remediation: 'Provide pause, stop, hide, disable, or adjustment controls as applicable.',
          verification:
            'Exercise the behavior in the browser and evaluate the complete timing policy.',
        }),
      );
    }
  }

  if (options.profile === 'cvi' || options.profile === 'all') manualChecks.push(...cviChecks());
  if (options.profile === 'switch' || options.profile === 'all') {
    manualChecks.push(...switchChecks());
  }

  return {
    schemaVersion: SCHEMA_VERSION,
    producer: {
      name: 'accessibility-devkit',
      version: RUNTIME_VERSION,
      runtime: options.runtime ?? 'node',
    },
    target: { kind: 'file', value: options.target },
    summary: summarize(findings, manualChecks),
    findings,
    manualChecks,
  };
}

File: python/src/accessibility_devkit/scanner.py
"""Conservative dependency-free HTML source scanner."""

import re

from .report import create_report, item


TAG_PATTERN = re.compile(r"<([a-zA-Z][\w:-]*)(\s[^<>]*?)?\s*/?>")
ATTR_PATTERN = re.compile(r"([^\s=/>]+)(?:\s*=\s*(?:\"([^\"]*)\"|'([^']*)'|([^\s>]+)))?")


def _attributes(raw):
    return {match.group(1).lower(): next((value for value in match.groups()[1:] if value is not None), "") for match in ATTR_PATTERN.finditer(raw)}


def _location(source, index, excerpt):
    line_start = source.rfind("\n", 0, index) + 1
    return {
        "line": source.count("\n", 0, index) + 1,
        "column": index - line_start + 1,
        "excerpt": re.sub(r"\s+", " ", excerpt).strip()[:160],
    }


def _tags(source):
    return [
        {
            "name": match.group(1).lower(),
            "attributes": _attributes(match.group(2) or ""),
            "index": match.start(),
            "location": _location(source, match.start(), match.group(0)),
        }
        for match in TAG_PATTERN.finditer(source)
    ]


def _manual_baseline():
    return [
        item(
            "color-only-communication",
            "Confirm that color is not the only way information, status, or required action is conveyed.",
            certainty="manual",
            severity="info",
            evidence="human",
            wcag=["1.4.1"],
            remediation="Pair color with text, shape, pattern, position, or another programmatically available cue.",
            verification="Review each state visually and with color or custom styles disabled.",
        ),
        item(
            "contrast-rendered",
            "Measure rendered foreground and background colors, including states, images, gradients, and transparency.",
            certainty="manual",
            severity="info",
            evidence="runtime",
            wcag=["1.4.3", "1.4.11"],
            remediation="Adjust rendered colors that do not meet the applicable text or non-text threshold.",
            verification="Measure computed colors in the browser and inspect every interactive state.",
        ),
    ]


def _profile_checks(profile):
    checks = []
    if profile in ("cvi", "all"):
        cvi = [
            ("cvi-individual-profile", "Review the experience against the person’s individual CVI profile and preferred presentation."),
            ("cvi-visual-complexity", "Review clutter, crowding, background complexity, and competing visual targets."),
            ("cvi-fatigue-motion", "Review visual fatigue, latency, motion, and the time needed to find and interpret targets."),
            ("cvi-multisensory-alternatives", "Confirm that important information has usable nonvisual or multisensory alternatives."),
        ]
        for rule_id, message in cvi:
            checks.append(item(rule_id, message, classification="supplemental", certainty="manual", severity="info", evidence="human", remediation="Adapt the presentation to the person’s documented needs and preferences.", verification="Evaluate with the person or a qualified specialist; no universal palette or ratio establishes CVI access."))
    if profile in ("switch", "all"):
        switch = [
            ("switch-control-verification", "Complete the primary flow with the operating system’s Switch Control.", []),
            ("switch-timing-preferences", "Confirm scan speed, dwell, repeat filtering, and time limits can match the person’s needs.", []),
            ("switch-simple-action", "Confirm complex pointer gestures have a single-switch or simple-action alternative.", ["2.5.1", "2.5.7"]),
        ]
        for rule_id, message, wcag in switch:
            checks.append(item(rule_id, message, classification="supplemental", certainty="manual", severity="info", evidence="human", wcag=wcag, remediation="Use native semantics, simple actions, and adjustable timing.", verification="Complete the task with the person’s switch configuration."))
    return checks


def scan_source(source, target, profile=None):
    tags = _tags(source)
    findings = []
    manual = _manual_baseline()
    html = next((tag for tag in tags if tag["name"] == "html"), None)
    if not html or not html["attributes"].get("lang", "").strip():
        findings.append(item("document-language", "The document does not declare a default language.", wcag=["3.1.1"], location=html["location"] if html else {"line": 1, "column": 1}, remediation="Add a valid lang attribute to the html element.", verification="Inspect the accessibility tree and confirm language changes are also marked."))

    ids = {}
    for tag in tags:
        identifier = tag["attributes"].get("id")
        if not identifier:
            continue
        if identifier in ids:
            findings.append(item("duplicate-id", f"The id {identifier!r} is used more than once.", wcag=["1.3.1", "4.1.2"], location=tag["location"], remediation="Give each relationship target a unique id and update its references.", verification=f"Confirm every reference to {identifier!r} resolves to one intended element."))
        else:
            ids[identifier] = tag

    for tag in (candidate for candidate in tags if candidate["name"] == "img"):
        attrs = tag["attributes"]
        if "alt" not in attrs:
            findings.append(item("image-alt-missing", "An image has no alt attribute.", wcag=["1.1.1"], location=tag["location"], remediation='Add contextual alternative text, or alt="" when the image is intentionally decorative.', verification="Review the image in context and confirm the accessible name communicates its purpose."))
        elif not attrs["alt"].strip():
            manual.append(item("image-alt-empty-context", "An image uses empty alternative text; that can be correct only when its content is redundant or decorative.", certainty="manual", severity="info", evidence="human", wcag=["1.1.1"], location=tag["location"], remediation='Keep alt="" for decorative images; otherwise provide concise contextual alternative text.', verification="Review adjacent content and the task with images unavailable."))

    label_fors = {tag["attributes"]["for"] for tag in tags if tag["name"] == "label" and tag["attributes"].get("for")}
    for tag in (candidate for candidate in tags if candidate["name"] in ("input", "select", "textarea")):
        attrs = tag["attributes"]
        if attrs.get("type", "text").lower() in ("hidden", "button", "submit", "reset", "image"):
            continue
        named = (attrs.get("id") in label_fors) or any(attrs.get(key, "").strip() for key in ("aria-label", "aria-labelledby", "title"))
        if not named:
            findings.append(item("form-label", "A form control has no source-visible label or accessible-name reference.", certainty="potential", wcag=["1.3.1", "3.3.2", "4.1.2"], location=tag["location"], remediation="Associate a visible label with the control and preserve its accessible name.", verification="Inspect the computed accessible name; account for labels created at runtime."))

    for tag in tags:
        tabindex = tag["attributes"].get("tabindex")
        try:
            positive = tabindex is not None and int(tabindex) > 0
        except ValueError:
            positive = False
        if positive:
            findings.append(item("focus-positive-tabindex", "A positive tabindex can create a focus order that diverges from reading order.", certainty="potential", severity="warning", wcag=["2.4.3"], location=tag["location"], remediation='Use DOM order and tabindex="0" or native focusability instead of a positive value.', verification="Traverse the complete flow with Tab, Shift+Tab, and Switch Control."))

    previous_heading = 0
    for tag in (candidate for candidate in tags if re.fullmatch(r"h[1-6]", candidate["name"])):
        level = int(tag["name"][1])
        if previous_heading and level > previous_heading + 1:
            findings.append(item("heading-order", f"Heading level jumps from h{previous_heading} to h{level}.", classification="advisory", certainty="potential", severity="warning", wcag=["1.3.1", "2.4.6"], location=tag["location"], remediation="Use heading levels that reflect the content hierarchy without skipping a parent level.", verification="Review the heading outline and confirm it matches the information structure."))
        previous_heading = level

    for match in re.finditer(r"<a\b[^>]*>([\s\S]*?)</a>", source, re.IGNORECASE):
        text = re.sub(r"\s+", " ", re.sub(r"<[^>]+>", " ", match.group(1))).strip()
        if not text or re.fullmatch(r"click here|read more|learn more|more", text, re.IGNORECASE):
            findings.append(item("link-purpose", "A link has empty or generic source-visible text.", certainty="potential", severity="warning", wcag=["2.4.4"], location=_location(source, match.start(), match.group(0)[:120]), remediation="Write link text that identifies the destination or action in context.", verification="Review the computed accessible name and a links-only list."))

    for match in re.finditer(r"<table\b[^>]*>([\s\S]*?)</table>", source, re.IGNORECASE):
        if not re.search(r"<th\b", match.group(1), re.IGNORECASE):
            manual.append(item("table-headers-context", "A table has no source-visible header cells; determine whether it presents data or is only layout.", certainty="manual", severity="info", evidence="human", wcag=["1.3.1"], location=_location(source, match.start(), match.group(0)[:120]), remediation="For data tables, add correctly scoped headers and a useful caption when needed.", verification="Navigate the table by row and column with a screen reader."))

    for tag in tags:
        attrs = tag["attributes"]
        interactive = tag["name"] in ("button", "input", "select", "textarea") or (tag["name"] == "a" and bool(attrs.get("href"))) or bool(re.fullmatch(r"button|link|checkbox|radio|switch|tab|menuitem", attrs.get("role", "")))
        if not interactive:
            continue
        style = attrs.get("style", "")
        width = re.search(r"(?:^|;)\s*width\s*:\s*([\d.]+)px", style, re.IGNORECASE)
        height = re.search(r"(?:^|;)\s*height\s*:\s*([\d.]+)px", style, re.IGNORECASE)
        if (width and float(width.group(1)) < 24) or (height and float(height.group(1)) < 24):
            manual.append(item("target-size-spacing", "A source dimension is below 24 CSS pixels; rendered size, spacing, and WCAG exceptions still require review.", certainty="manual", severity="info", evidence="runtime", wcag=["2.5.8"], location=tag["location"], remediation="Increase the rendered target or provide sufficient spacing or an applicable exception.", verification="Measure rendered boxes and spacing at representative viewport sizes and zoom levels."))

    for tag in tags:
        attrs = tag["attributes"]
        executable = (tag["name"] in ("audio", "video") and "autoplay" in attrs) or (tag["name"] == "meta" and attrs.get("http-equiv", "").lower() == "refresh")
        if executable:
            manual.append(item("timing-media-control", "Source markup configures automatic media or refresh behavior that needs runtime timing review.", certainty="manual", severity="info", evidence="runtime", wcag=["2.2.1", "2.2.2"], location=tag["location"], remediation="Provide pause, stop, hide, disable, or adjustment controls as applicable.", verification="Exercise the behavior in the browser and evaluate the complete timing policy."))

    manual.extend(_profile_checks(profile))
    return create_report("file", target, findings, manual)
